libs/execution_context/ebpf_program.c

```diff
@@ -1863,38 +2005,55 @@ _ebpf_program_compute_program_information_hash(
     ebpf_helper_id_to_index_t* helper_id_to_index = NULL;
     const ebpf_program_info_t* type_specific_program_info = type_specific_program_information_data->program_info;
     const ebpf_program_info_t* general_program_info = general_program_information_data->program_info;
-    uint32_t helper_function_count = type_specific_program_info->count_of_program_type_specific_helpers +
-                                     general_program_info->count_of_program_type_specific_helpers;
+    size_t helper_function_count = count_of_actual_helper_ids;
     uint32_t helper_function_index = 0;
 
-    helper_id_to_index = (ebpf_helper_id_to_index_t*)ebpf_allocate_with_tag(
-        helper_function_count * sizeof(ebpf_helper_id_to_index_t), EBPF_POOL_TAG_PROGRAM);
-    if (helper_id_to_index == NULL) {
-        result = EBPF_NO_MEMORY;
-        goto Exit;
-    }
+    // It is possible that a program does not invoke any helper function.
+    if (helper_function_count) {
+        helper_id_to_index = (ebpf_helper_id_to_index_t*)ebpf_allocate_with_tag(
+            helper_function_count * sizeof(ebpf_helper_id_to_index_t), EBPF_POOL_TAG_PROGRAM);
+        if (helper_id_to_index == NULL) {
+            result = EBPF_NO_MEMORY;
+            goto Exit;
+        }
 
-    // Copy global helpers to helper_id_to_index.
-    for (uint32_t index = 0; index < general_program_info->count_of_program_type_specific_helpers; index++) {
-        helper_id_to_index[helper_function_index].helper_id =
-            general_program_info->program_type_specific_helper_prototype[index].helper_id;
-        helper_id_to_index[helper_function_index].helper_function_prototype =
-            &general_program_info->program_type_specific_helper_prototype[index];
-        helper_function_index++;
-    }
+        // Copy global helpers to helper_id_to_index.
+        for (uint32_t index = 0; index < general_program_info->count_of_program_type_specific_helpers; index++) {
+            uint32_t helper_id = general_program_info->program_type_specific_helper_prototype[index].helper_id;
+            if (!_ebpf_contains_helper_id(actual_helper_ids, count_of_actual_helper_ids, helper_id)) {
```

Comment:
Is this (and the below) expensive in practice? Is there any way we can optimize this a bit? (I see below sorts the output for example. Is actual_helper_ids sorted at all?)

Reply:
Yes we could ideally try to optimize it. But this is in control path when the program is being loaded. `actual_helper_ids` may or may not be sorted today (for JIT modules, it is sorted, but for native, it is not).